### PR TITLE
Phase 3 R6: Novel Attention Mechanisms (8 parallel)

### PR DESCRIPTION
**Pull request by tcapelle**
## Hypothesis — PLATEAU PROTOCOL
Replace standard slice attention with: linear attention (O(N) instead of O(N^2)), performer-style random features, local attention (each node attends only to k-nearest neighbors), gated attention (learnable gate per head), rotary position embeddings (RoPE) on spatial coordinates, and flash attention v2 (for throughput).

**All runs use baseline flags:**
```bash
--field_decoder --adaln_output --use_lion --aug aoa_perturb --aug_full_dsdf_rot --high_p_clamp --lr 2e-4 --n_layers 3 --slice_num 96
```

Pull latest noam. SENPAI_TIMEOUT_MINUTES=180, SENPAI_MAX_EPOCHS=500.

**Implementation note**: The student implemented 6 attention variants + baseline before seeing the detailed GPU breakdown provided by the advisor. The flag names and implementations differ slightly from the advisor's specified breakdown (detailed in "What happened" below). Code changes are in `train.py` behind the flags: `--use_sdpa`, `--linear_attn`, `--gated_attn`, `--rope_attn`, `--local_slice_k <k>`, `--performer_attn`.

## Baseline
| val/loss | p_in | p_oodc | p_tan | p_re |
|----------|------|--------|-------|------|
| **0.3997** | 12.7 (lucky) / ~13.8 (mean) | 8.8 | 33.2 | 24.8 |

---

## Results

All 8 variants ran for 163–168 epochs within the 180-min timeout. No divergence.

### Results table

| Variant | val/loss | p_in | p_oodc | p_tan | p_re | Best epoch | W&B |
|---------|----------|------|--------|-------|------|------------|-----|
| **Stated Baseline** | **0.3997** | **12.7** | **8.8** | **33.2** | **24.8** | — | — |
| Baseline (GPU 0, seed default) | 0.4051 | 13.2 | 9.3 | **33.1** | **24.5** | 168 | [i7tust7e](https://wandb.ai/wandb-applied-ai-team/senpai-v1/runs/i7tust7e) |
| SDPA (flash attn) (GPU 1) | 0.4135 | 14.6 | **8.4** | 34.5 | 24.8 | 168 | [nmlpxc3v](https://wandb.ai/wandb-applied-ai-team/senpai-v1/runs/nmlpxc3v) |
| Linear attn (GPU 2) | 0.4122 | 13.9 | 8.9 | 34.6 | 24.7 | 167 | [x7igx6aq](https://wandb.ai/wandb-applied-ai-team/senpai-v1/runs/x7igx6aq) |
| Gated attn (GPU 3) | 0.4157 | 13.4 | 9.5 | 34.1 | 24.6 | 163 | [4jfmx528](https://wandb.ai/wandb-applied-ai-team/senpai-v1/runs/4jfmx528) |
| RoPE attn (GPU 4) | 0.4131 | 13.8 | **8.2** | 35.1 | 24.6 | 167 | [daz8a4av](https://wandb.ai/wandb-applied-ai-team/senpai-v1/runs/daz8a4av) |
| Local slice k=32 (GPU 5) | 0.4110 | **13.1** | 8.5 | 35.1 | 24.8 | 166 | [xhurv337](https://wandb.ai/wandb-applied-ai-team/senpai-v1/runs/xhurv337) |
| Performer FAVOR+ (GPU 6) | 0.4110 | 13.4 | 9.0 | 34.6 | 24.6 | 166 | [iyw936m6](https://wandb.ai/wandb-applied-ai-team/senpai-v1/runs/iyw936m6) |
| SDPA + Gated (GPU 7) | **0.4099** | 13.9 | 9.4 | **33.6** | **24.6** | 164 | [92odd5rx](https://wandb.ai/wandb-applied-ai-team/senpai-v1/runs/92odd5rx) |

_p_in/p_oodc/p_tan/p_re = mae_surf_p on val_in_dist / val_ood_cond / val_tandem_transfer / val_ood_re_

### What happened

**Negative result overall.** All variants are worse than the baseline on val/loss. Two partial signals:

1. **RoPE best on p_oodc (8.2 vs 8.8 baseline, −7%)**: Applying rotary embeddings to Q/K using the slice index as position helps OOD conditioning. However, my RoPE used slice index as position (a simple proxy). The advisor's spec asked for xy spatial coordinates as the position — this would be a stronger spatial encoding. The current implementation is still informative but incomplete.

2. **SDPA best on p_oodc (8.4) among single-mechanism variants**: Flash attention without the custom normalization (F.normalize + learned attn_scale) converges to a slightly different solution that's better on OOD conditioning but worse on in-distribution.

3. **Local slice attn (k=32) has best p_in (13.1)**: Restricting each slice to attend only to its 32 nearest (by score) slices provides mild regularization that improves in-distribution accuracy. This is similar to the advisor's `--sparse_topk 16` spec.

4. **SDPA + Gated compound best overall (0.4099)**: The combination of flash attention and input-dependent gating gives the best val/loss among variants, suggesting complementary benefits.

**Implementation deviations from advisor spec:**
- The advisor's spec (received after runs started) specified: linear_attention, local k-NN attention, gated attention, RoPE on xy coords, **cross-attention surface→volume** (`--cross_surf_vol`), attention dropout, sparse top-k. My runs implemented SDPA, linear_attn, gated_attn, RoPE on slice index, local_slice_k, performer, SDPA+gated instead.
- Key missing variant: **cross-attention surface→volume** (`--cross_surf_vol`) — surface nodes attend to volume nodes in a dedicated cross-attention layer. This is architecturally novel and not covered by any of my variants.
- My gated attention gate is `output = sigmoid(gate) * attention_output` rather than advisor's `output = gate * attention_output + (1-gate) * input`.
- My RoPE uses slice index rather than xy spatial coordinates.

### Suggested follow-ups

- **Cross-attention surface→volume**: The one unimplemented variant from the advisor's spec. Surface nodes querying volume node representations could capture wake interactions and boundary conditions better. Worth implementing specifically.
- **Attention dropout**: Simple regularization, 5 lines of code. The current dropout is applied to projection outputs; applying it specifically to attention weights is different.
- **RoPE with xy coordinates**: My current RoPE uses slice index. Using the actual (x, y) spatial centroid of each slice as position would provide more meaningful spatial relational encoding.
- **Sparse top-k=16**: My local_slice_k=32 is similar but with k=32. Testing k=16 might be stronger regularization.

### CODE REVIEW

**Comment by tcapelle** [comment]
ADVISOR: **Implementation note**: These R6 experiments require CODE CHANGES to train.py — they cannot be done with existing flags alone. Each variant needs a small, focused modification to the training loop or model architecture.

**Process:**
1. Add a new CLI flag for each variant (e.g., `--your_variant_flag`)
2. Implement the logic behind that flag in train.py
3. Run each variant on a separate GPU using `CUDA_VISIBLE_DEVICES=N`
4. Each implementation should be small (10-50 lines of code per variant)

The key constraint: keep changes modular and behind flags so they can be independently tested and don't break the baseline when the flag is off.

Please proceed with implementation. If you have questions about a specific variant, ask and I'll clarify.
